example/printer_raw.dart

```diff
@@ -0,0 +1,160 @@
+      pDocInfo.ref.pDocName = printerName.toNativeUtf16();
+      // We can use RAW or TEXT data types
+      pDocInfo.ref.pDatatype = TEXT(dataType); // RAW, TEXT or XPS_PASS
+      // "To print to a printer, set this to NULL."
+      pDocInfo.ref.pOutputFile = nullptr;
+
+      //https://docs.microsoft.com/en-us/windows/win32/printdocs/startdocprinter
+      fSuccess = StartDocPrinter(
+          phPrinter.value,
+          // "The version of the structure to which pDocInfo points. This value must be 1."
+          1,
+          pDocInfo);
+
+      if (fSuccess == 0) {
+        final error = GetLastError();
+        throw Exception(
+            'StartDocPrinter error, status: $fSuccess, error: $error');
+      }
+
+      return phPrinter;
+    } finally {
+      free(pPrinterName);
+      free(pDocInfo);
+    }
+  }
+
+  bool _startRawPrintPage(Pointer<HANDLE> phPrinter) {
+    //https://docs.microsoft.com/en-us/windows/win32/printdocs/startpageprinter
+    return StartPagePrinter(phPrinter.value) > 0;
+  }
+
+  bool _endRawPrintPage(Pointer<HANDLE> phPrinter) {
+    return EndPagePrinter(phPrinter.value) > 0;
+  }
+
+  bool _endRawPrintJob(Pointer<HANDLE> phPrinter) {
+    return EndDocPrinter(phPrinter.value) > 0 &&
+        ClosePrinter(phPrinter.value) > 0;
+  }
+
+  bool _printRawData(Pointer<HANDLE> phPrinter, String dataToPrint) {
+    final cWritten = calloc<DWORD>();
+    // Convert the Dart String to a C++ compatible String.
+    final data = _toCString(dataToPrint);
+
+    try {
+      // https://docs.microsoft.com/en-us/windows/win32/printdocs/writeprinter
+      final result =
+          WritePrinter(phPrinter.value, data, dataToPrint.length, cWritten);
+
+      if (dataToPrint.length != cWritten.value) {
+        final error = GetLastError();
+        throw Exception('WritePrinter error, status: $result, error: $error');
+      }
+
+      return result > 0;
+    } finally {
+      free(cWritten);
+      free(data);
+    }
+  }
+
+  bool printLines(List<String> data) {
+    // Example taken from:
+    // https://docs.microsoft.com/en-us/windows/win32/printdocs/sending-data-directly-to-a-printer
+    Pointer<HANDLE> printerHandle;
+    var res = false;
+
+    if (data.isEmpty) {
+      return res;
+    }
+
+    printerHandle = _startRawPrintJob(
+        printerName: printerName,
+        documentTitle: 'My document',
+        dataType: 'RAW');
+
+    try {
+      res = _startRawPrintPage(printerHandle);
+
+      for (final item in data) {
+        if (res) {
+          res = _printRawData(printerHandle, item);
+        }
+      }
+    } finally {
+      _endRawPrintPage(printerHandle);
+      _endRawPrintJob(printerHandle);
+      free(printerHandle);
+    }
+
+    return res;
+  }
+}
+
+void main() {
+  // Example: ESC/POS sequence to open the cash drawer
+  const openCashDrawer = '\x1b\x70\x00';
+
+  // NOTE: You can get the printer name from the printer_list.dart example
```

Comment:
By using Arena to manage allocations, you can have Dart automatically free memory when RawPrinter falls out of scope.

Reply:
TIL again, thanks.